File: app/src/main/java/com/google/android/stardroid/activities/ImageGalleryActivity.java
// Copyright 2010 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.android.stardroid.activities;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.BaseAdapter;
import android.widget.Gallery;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.preference.PreferenceManager;

import com.google.android.stardroid.R;
import com.google.android.stardroid.activities.util.ActivityLightLevelChanger;
import com.google.android.stardroid.activities.util.ActivityLightLevelManager;
import com.google.android.stardroid.gallery.GalleryFactory;
import com.google.android.stardroid.gallery.GalleryImage;
import com.google.android.stardroid.util.Analytics;
import com.google.android.stardroid.util.MiscUtil;

import java.util.List;

import javax.inject.Inject;

/**
 * Displays a series of images to the user.  Selecting an image
 * invokes Sky Map Search.
 *
 * @author John Taylor
 */
public class ImageGalleryActivity extends InjectableActivity {
  /** The index of the image id Intent extra.*/
  public static final String IMAGE_ID = "image_id";

  private static final String TAG = MiscUtil.getTag(ImageGalleryActivity.class);
  private List<GalleryImage> galleryImages;

  private ActivityLightLevelManager activityLightLevelManager;
  @Inject
  Analytics analytics;

  private class ImageAdapter extends BaseAdapter {
    public int getCount() {
      return galleryImages.size();
    }

    public Object getItem(int position) {
      return position;
    }
    public long getItemId(int position) {
      return position;
    }

    /**
     * Returns a new ImageView to be displayed, depending on the position passed.
     */
    public View getView(int position, View convertView, ViewGroup parent) {
      Log.d(TAG, "Get view called for position "+ position);
      ViewGroup imagePanel;
      if (convertView != null && convertView instanceof ViewGroup) {
        imagePanel = (ViewGroup) convertView;
      } else {
        imagePanel = (ViewGroup) getLayoutInflater().inflate(
            R.layout.imagedisplaypanel, parent, false);
      }
      GalleryImage galleryImage = galleryImages.get(position);
      ImageView imageView = (ImageView) imagePanel.findViewById(R.id.image_gallery_image);
      imageView.setImageResource(galleryImage.imageId);
      TextView imageLabel = (TextView) imagePanel.findViewById(R.id.image_gallery_title);
      imageLabel.setText(galleryImage.name);
      return imagePanel;
    }
  }

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    getApplicationComponent().inject(this);
    setContentView(R.layout.imagegallery);
    activityLightLevelManager = new ActivityLightLevelManager(
        new ActivityLightLevelChanger(this, null),
        PreferenceManager.getDefaultSharedPreferences(this));
    this.galleryImages = GalleryFactory.getGallery(getResources()).getGalleryImages();
    addImagesToGallery();
  }

  @Override
  public void onStart() {
    super.onStart();
  }

  @Override
  public void onResume() {
    super.onResume();
    activityLightLevelManager.onResume();
  }

  @Override
  public void onPause() {
    super.onPause();
    activityLightLevelManager.onPause();
  }

  private void addImagesToGallery() {
    Gallery gallery = (Gallery) findViewById(R.id.image_gallery);
    ImageAdapter imageAdapter = new ImageAdapter();
    gallery.setAdapter(imageAdapter);
    gallery.setOnItemClickListener(new OnItemClickListener() {
      @Override
      public void onItemClick(AdapterView<?> parent, View v, int position, long id) {
        showImage(position);
      }
  });
  }

  /**
   * Starts the display image activity, and overrides the transition animation.
   */
  private void showImage(int position) {
    Intent intent = new Intent(ImageGalleryActivity.this, ImageDisplayActivity.class);
    intent.putExtra(ImageGalleryActivity.IMAGE_ID, position);
    startActivity(intent);
    overridePendingTransition(R.anim.fadein, R.anim.fastzoom);
  }
}

File: app/src/main/java/com/google/android/stardroid/activities/ImageGalleryActivity.kt
package com.google.android.stardroid.activities

import android.app.SearchManager
import android.content.Intent
import android.os.Bundle
import androidx.fragment.app.FragmentActivity
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.stardroid.R
import com.google.android.stardroid.activities.dialogs.ObjectInfoDialogFragment
import com.google.android.stardroid.activities.util.ActivityLightLevelChanger
import com.google.android.stardroid.activities.util.ActivityLightLevelManager
import com.google.android.stardroid.activities.util.NightModeHelper
import com.google.android.stardroid.education.ObjectInfo
import com.google.android.stardroid.education.ObjectInfoRegistry
import com.google.android.stardroid.gallery.GalleryAdapter
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

/**
 * Displays a scrollable grid of celestial object thumbnails sourced from [ObjectInfoRegistry].
 * Tapping a thumbnail shows its info card via [ObjectInfoDialogFragment].
 */
@AndroidEntryPoint
class ImageGalleryActivity : FragmentActivity(),
    ActivityLightLevelChanger.NightModeable,
    ObjectInfoDialogFragment.OnFindClickedListener {

    @Inject lateinit var registry: ObjectInfoRegistry
    @Inject lateinit var activityLightLevelManager: ActivityLightLevelManager

    private lateinit var galleryAdapter: GalleryAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_image_gallery)

        val items = registry.getAllWithImages()
        val recyclerView = findViewById<RecyclerView>(R.id.gallery_grid)
        recyclerView.layoutManager = GridLayoutManager(this, 3)
        galleryAdapter = GalleryAdapter(items) { info ->
            if (!supportFragmentManager.isStateSaved) {
                ObjectInfoDialogFragment.newInstance(info)
                    .show(supportFragmentManager, "ObjectInfo")
            }
        }
        recyclerView.adapter = galleryAdapter
    }

    override fun onResume() {
        super.onResume()
        activityLightLevelManager.onResume()
    }

    override fun onPause() {
        super.onPause()
        activityLightLevelManager.onPause()
    }

    override fun setNightMode(nightMode: Boolean) {
        NightModeHelper.applyActionBarNightMode(actionBar, this, nightMode)
        galleryAdapter.notifyDataSetChanged()
    }

    override fun onFindClicked(info: ObjectInfo) {
        // Ensure the relevant layers are visible so the object can be found.
        val prefs = PreferenceManager.getDefaultSharedPreferences(this)
        prefs.edit()
            .putBoolean("source_provider.0", true) // Stars
            .putBoolean("source_provider.1", true) // Constellations
            .putBoolean("source_provider.2", true) // Deep Sky Objects
            .putBoolean("source_provider.3", true) // Planets
            .apply()
        val searchIntent = Intent(this, DynamicStarMapActivity::class.java).apply {
            action = Intent.ACTION_SEARCH
            putExtra(SearchManager.QUERY, info.name)
        }
        startActivity(searchIntent)
    }
}
